app/controllers/schools/add_participants_controller.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+module Schools
+  class AddParticipantsController < ::Schools::BaseController
+    FORM_SESSION_KEY = :add_participant_form
+    FORM_PARAM_KEY = :schools_add_participant_form
+
+    skip_after_action :verify_authorized
+    before_action :ensure_form_present, except: :start
+    before_action :set_school_cohort
+
+    helper_method :add_participant_form
+
+    def start
+      session.delete(FORM_SESSION_KEY)
+      session[FORM_SESSION_KEY] = {
+        school_cohort_id: @school_cohort.id,
+      }
+      redirect_to action: :show, step: :type
+    end
+
+    def show
+      render current_step
+    end
+
+    def update
+      if add_participant_form.valid?(current_step)
+        add_participant_form.record_completed_step current_step
+        store_form_in_session
+        redirect_to action: :show, step: step_param(add_participant_form.next_step(current_step))
+      else
+        render current_step
+      end
+    end
+
+    def complete
+      @participant = add_participant_form.save!
+      @type = add_participant_form.type
+
+      session.delete(FORM_SESSION_KEY)
+    end
+
+  private
+
+    def add_participant_form
+      return @add_participant_form if defined?(@add_participant_form)
+
+      @add_participant_form = AddParticipantForm.new(session[FORM_SESSION_KEY])
+      @add_participant_form.assign_attributes(add_participant_form_params) if params[FORM_PARAM_KEY]
+
+      @add_participant_form
+    end
+
+    def store_form_in_session
+      session[FORM_SESSION_KEY] = add_participant_form.attributes
+    end
+
+    def current_step
+      params[:step].tr("-", "_").to_sym
+    end
+
+    def step_param(step)
+      step.to_s.tr("_", "-")
+    end
+
+    def back_link_path
+      previous_step = add_participant_form.previous_step(current_step)
+      return schools_cohort_participants_path unless previous_step
+
+      { action: :show, step: step_param(previous_step) }
+    end
```

Comment:
Under the hood, all the url_helpers exists only to generate a hash that is then passed to `url_for`. The minimal url hash requires at least controller and action key, with any additional params defined in the routes. Additionally, controller keeps the hash for the current route in `url_options` and uses this as a base for any future route generations, so that we only need to specify the keys that are different from the current route. Hash can be used wherever one would use the url_helper.

BTW, this is also why in the context of any schools/base controller we can freely us `school_cohort_path` instead of `school_cohort_path(@cohort.start_year)`. Rails will automatically use the cohort data from the current route.